File: api/models/workspace.py
# SPDX-FileCopyrightText: 2026 Brave Bear Studios
# SPDX-License-Identifier: AGPL-3.0-or-later
"""Workspace models (PLAT-09).

Mirror the ``workspaces`` table (tech-spec §7.1). Fields are snake_case in
Python and serialize to camelCase JSON via :class:`CamelModel`.
"""

from typing import Literal

from models.base import CamelModel

WorkspaceStatus = Literal["creating", "running", "stopped", "error", "destroyed"]


class Workspace(CamelModel):
    """A worker workspace row (tech-spec §7.1)."""

    id: str
    name: str
    status: WorkspaceStatus
    vmid: int | None
    node: str
    lxc_ip: str | None
    project_repo: str
    project_branch: str
    plugin_set: str
    created_at: str
    stopped_at: str | None
    destroyed_at: str | None
    deleted_at: str | None
    persistent: bool = False  # WSX-02; stored INTEGER 0/1, Pydantic coerces to bool


class WorkspaceCreate(CamelModel):
    """Request body for creating a workspace.

    ``node`` is OPTIONAL: ``None``/omitted signals AUTO placement (the saga
    selects the least-loaded fitting node via ``WorkspaceService.selectNode``),
    while an explicit node string is the unchanged manual path (WSX-01).
    """

    name: str
    project_repo: str
    project_branch: str = "main"
    plugin_set: str = "default"
    node: str | None = None
    persistent: bool = False  # opt-in; default ephemeral (CONTEXT-locked WSX-02)

File: ui/src/types/workspace.ts
// SPDX-FileCopyrightText: 2026 Brave Bear Studios
// SPDX-License-Identifier: AGPL-3.0-or-later

// Frontend domain types. Field names are camelCase to match the backend's
// CamelModel JSON (api/models/workspace.py serialized via model_dump(by_alias=True)):
// snake_case columns (lxc_ip, project_repo, …) surface as lxcIp, projectRepo, ….

/** The workspace lifecycle status, 1:1 with the backend `WorkspaceStatus` enum. */
export type WorkspaceStatus =
	| "creating"
	| "running"
	| "stopped"
	| "error"
	| "destroyed";

/** A worker workspace row (mirrors api/models/workspace.py::Workspace). */
export interface Workspace {
	id: string;
	name: string;
	status: WorkspaceStatus;
	vmid: number | null;
	node: string;
	lxcIp: string | null;
	projectRepo: string;
	projectBranch: string;
	pluginSet: string;
	createdAt: string;
	stoppedAt: string | null;
	destroyedAt: string | null;
	deletedAt: string | null;
}

/** Request body for POST /api/v1/workspaces (mirrors WorkspaceCreate). */
export interface WorkspaceCreate {
	name: string;
	projectRepo: string;
	projectBranch?: string;
	pluginSet?: string;
	// Optional, mirroring the backend `node: str | None = None`: null/omitted means
	// "auto-select the least-loaded node"; a node string is an explicit manual pick.
	node?: string | null;
}

/** Per-node capacity from GET /api/v1/nodes (Plan 02-01 backend). */
export interface NodeCapacity {
	node: string;
	memoryUsedFraction: number | null;
	capacityThreshold: number;
	overThreshold: boolean;
}

/** The standard response envelope (mirrors api/lib/envelope.py). */
export interface ApiEnvelope<T> {
	data: T | null;
	meta: { requestId: string; timestamp: string };
	error: { code: string; message: string } | null;
}

/** Terminal connection state surfaced by useTerminal (TERM-06 overlay states). */
export type TerminalState =
	| "connecting"
	| "open"
	| "reconnecting"
	| "error"
	| "closed";
